pulser/simulation/simulation.py

```diff
@@ -15,116 +15,302 @@
+        diff_noise_set = new_noise_set - old_noise_set
+        param_dict: dict[str, Any] = asdict(self._config)
+        del param_dict["spam_dict"]
+        del param_dict["doppler_sigma"]
+        param_dict['noise'] = tuple(new_noise_set)
+        if 'SPAM' in diff_noise_set:
+            param_dict['eta'] = config.eta
+            param_dict['epsilon'] = config.epsilon
+            param_dict['epsilon_prime'] = config.epsilon_prime
+        if 'doppler' in diff_noise_set:
+            param_dict['temperature'] = config.temperature * 1.e6
+        if 'amplitude' in diff_noise_set:
+            param_dict['laser_waist'] = config.laser_waist
+        if 'dephasing' in diff_noise_set:
+            param_dict['dephasing_prob'] = config.dephasing_prob
+        self.set_config(SimConfig(**param_dict))
+
+    def show_config(self) -> None:
+        """Shows current configuration."""
+        print(self._config)
+
+    def reset_config(self) -> None:
+        """Resets configuration to default."""
+        self.set_config(SimConfig())
+        print("Configuration has been set to default.")
+
+    def _set_param_from_config(self) -> None:
+        """Sets all relevant Simulation parameters from its SimConfig.
+
+        Called every time a new configuration is loaded in order to update
+        Simulation parameters.
+        """
+        # Noise, samples and Hamiltonian update routine
+        self._update_hamiltonian()
+        if 'dephasing' in self.config.noise and not self._collapse_ops:
+            self._init_dephasing()
+
+    def _update_hamiltonian(self) -> None:
+        """Noise, samples and Hamiltonian update routine.
+
+        Updates:
+            -Noise parameters, randomly chosen at each run
+            -New sequence samples that may have changed due to noise
+            -Sequence Hamiltonian from samples.
+        """
+        self._update_noise()
+        self._reset_samples()
+        self._extract_samples()
+        self._construct_hamiltonian()
+
+    def _update_noise(self) -> None:
+        """Updates noise random parameters.
+
+        Used at the start of each run. If SPAM isn't in chosen noises, all
+            atoms are set to be correctly prepared.
+        """
+        if 'SPAM' in self.config.noise and self.config.eta > 0:
+            self._prepare_bad_atoms()
+        elif 'SPAM' not in self.config.noise:
+            # Tag True if atom is badly prepared
+            self._bad_atoms: dict[Union[str, int], bool] = {
+                qid: False for qid in self._qid_index}
+        if 'doppler' in self.config.noise:
+            self._prepare_doppler_detune()
+
+    def _prepare_bad_atoms(self) -> None:
+        """Chooses atoms that will be badly prepared."""
+        dist = (np.random.uniform(size=len(self._qid_index)) <
+                self.config.spam_dict['eta'])
+        self._bad_atoms = dict(zip(self._qid_index, dist))
+
+    def _prepare_doppler_detune(self) -> None:
```

Comment:
The same comment for `_prepare_bad_atoms` applies here. We want modularity, but it doesn't make sense to have these defined as method if they are only called once.